Rg/warm start option

grishabh147 opened the pull request:
This PR adds an option to warm-start the `FullSpaceModel` with initial primal and dual solution guesses.

This also adds the code to construct a `ProxAugmentedBlockNLPModel` which adds a proximal term to the augmented objective function.

grishabh147 committed:
added warm-start option

grishabh147 committed:
add ProxAugBlock

grishabh147 committed:
add ProxAugBlock

grishabh147 committed:
add warm-start option for FullSpaceModel

frapac commented on src/full_space.jl:
```diff
@@ -11,16 +11,24 @@ end
 # Arguments
 
 - `m::AbstractBlockNLPModel`: name of the BlockNLPModel
+- `x0::Union{Vector{Float64}, Nothing}`: warm start primal solution
```
Maybe specify that by default `x0` is a vector of zeroes?

grishabh147 replied:
I am already doing this through these two lines:
```Julia
isnothing(x0) && (x0 = [m.blocks[i].meta.x0 for i in 1:nb])
x0 = reduce(vcat, x0)

frapac replied:
I would suggest defining these two lines directly in a function `NLPModels.get_x0`

frapac commented on src/full_space.jl:
```diff
@@ -11,16 +11,24 @@ end
     FullSpaceModel(
-        ::Type{T}, m::AbstractBlockNLPModel
+        m::AbstractBlockNLPModel
+        x0::Union{Vector{Float64}, Nothing} = nothing, 
```
how about using directly?
```julia
x0 = NLPModels.get_x0(m)
```
and implements the corresponding `NLPModels.get_x0` function for the different `AbstractBlockNLPModel`?

frapac commented on src/prox_augmented_block.jl:
```diff
@@ -0,0 +1,248 @@
+    A::AbstractArray # linking matrix
+    b::AbstractVector
+    sol::AbstractVector # current primal solution
+    P::AbstractArray # proximal term penalty parameters
```
do we assume that the proximal parameter is a vector (so a diagonal matrix) by default?

grishabh147 replied:
No, `P` needs to be a matrix. Isn't `AbstractMatrix <: AbstractArray`? I can make it an `AbstractMatrix` is that's more clear.

frapac replied:
You are right that `AbstractMatrix <: AbstractArray`. I would suggest typing `P::AbstractMatrix`, to force it to be a matrix.

frapac commented on src/prox_augmented_block.jl:
```diff
@@ -0,0 +1,248 @@
+- `A::AbstractArray`: full linking matrix
+- `b::AbstractVector`: RHS vector for linking constraints
+- `sol::AbstractVector`: vector of primal variables
+- `P::Vector{AbstractArray}`: matrix of proximal term's penalty parameters
```
I am not sure to understand if `P` is an `AbstractArray` or a `Vector{AbstractArray}`?

grishabh147 replied:
Sorry, this is a typo.

frapac commented on src/prox_augmented_block.jl:
```diff
@@ -0,0 +1,248 @@
+    subproblem::AbstractNLPModel
+    λ::AbstractVector # dual variables
+    ρ::Number # penalty Parameter
+    A::AbstractArray # linking matrix
```
Maybe specify `AbstractMatrix` ?

frapac commented on src/prox_augmented_block.jl:
```diff
@@ -0,0 +1,248 @@
+        lcon = nlp.meta.lcon,
+        ucon = nlp.meta.ucon,
+        minimize = true,
+        name = "augmented_block",
```
maybe change the name to `"proximal_augmented_block"`?

frapac commented on src/prox_augmented_block.jl:
```diff
@@ -0,0 +1,248 @@
+- `y::Union{AbstractVector, Nothing}`: vector of dual variables
+- `vals::Union{AbstractVector, Nothing}`: nonzero values of the Hessian matrix
+"""
+function get_hessian(
```
naive question: can we reuse the function we have implemented previously for `AugmentedBlockNLPModels`?

frapac commented on src/prox_augmented_block.jl:
```diff
@@ -0,0 +1,248 @@
+    local_sol[nlp.subproblem.var_idx] = x
+
+    grad!(nlp.subproblem.problem_block, x, g)
+    g .+=
```
I would suggest using `mul!` (to dispatch on BLAS operations) instead of the broadcast operator here

frapac commented on .vscode/settings.json:
```diff
@@ -0,0 +1 @@
+{}
```
do we have to version this file?

frapac reviewed (changes requested):
Nice! I only have a few minor comments. 

The main question is what do we do with `get_hessian`? I think we can use the same function both for `AugmentedNLPBlockModel` and for `ProxAugmentedNLPBlockModel`. But I would suggest addressing this in a future PR.

grishabh147 committed:
address PR comments

grishabh147 committed:
remove .vscode

grishabh147 commented:
@frapac This is now ready for a review. In addition to addressing your comments from last week, I have also tried to make changes to `augmented_block.jl` code to make the hessian computation step more efficient. Let me know if you have any comments/suggestions on my implementation.

frapac reviewed (approved):
Thanks for the update! I think this PR is ready to merge
